File: src/sim/Simulation.cpp
#include "Simulation.hpp"
#include "DecisionTree.hpp"
#include "SpatialGrid.hpp"
#include <algorithm>
#include <cmath>
#include <glaze/glaze.hpp>
#include <random>
#include <ranges>

// ── Glaze wire-format structs ─────────────────────────────────────────────────
// Must have external linkage (no anonymous namespace) for glaze reflection.

struct DroneSnap {
    int              id{}, team{};
    double           x{}, y{}, z{};
    double           vx{}, vy{}, vz{};
    double           health{}, maxHp{};
    std::string_view state{};
};

struct WorldSnap {
    std::uint64_t          tick{};
    std::string_view       phase{};
    std::size_t            alphaAlive{}, betaAlive{};
    double                 resetIn{};
    std::vector<DroneSnap> drones{};
};

// ── Module-internal helpers (anonymous namespace) ─────────────────────────────

namespace {

constexpr double kTickDt      = 0.05;
constexpr double kMaxSpeed    = 9.0;
constexpr double kPatrolSpeed = 4.5;
constexpr double kBoundsHalf  = 150.0;
constexpr double kBoundsYMax  = 100.0;
constexpr double kSteerFactor = 0.12;
constexpr double kResetDelay  = 5.0;

thread_local std::mt19937 rng{std::random_device{}()};

double rand_range(double lo, double hi) {
    return std::uniform_real_distribution{lo, hi}(rng);
}
Vec3 random_waypoint() {
    return { rand_range(-100, 100), rand_range(10, 70), rand_range(-100, 100) };
}
void clamp_pos(Vec3& v) noexcept {
    v.x = std::clamp(v.x, -kBoundsHalf, kBoundsHalf);
    v.y = std::clamp(v.y, 2.0,          kBoundsYMax);
    v.z = std::clamp(v.z, -kBoundsHalf, kBoundsHalf);
}

} // anonymous namespace

// ── Construction ──────────────────────────────────────────────────────────────

Simulation::Simulation(std::size_t drones_per_team)
    : drones_per_team_{drones_per_team}
{
    drones_.reserve(drones_per_team * 2);
    decisions_.reserve(drones_per_team * 2);
    hits_.reserve(drones_per_team * 2);
    json_buf_.reserve(64 * 1024);
    spawn_drones();
}

// ── Public API ────────────────────────────────────────────────────────────────

void Simulation::step()
{
    std::scoped_lock lock{mu_};
    ++tick_;

    if (phase_ == Phase::Running) {
        if (alive_count(Team::Alpha) == 0 || alive_count(Team::Beta) == 0) {
            phase_           = Phase::Resetting;
            reset_countdown_ = kResetDelay;
        }
    }

    if (phase_ == Phase::Resetting) {
        reset_countdown_ -= kTickDt;
        if (reset_countdown_ <= 0.0) {
            spawn_drones();
            phase_ = Phase::Running;
        }
        return;
    }

    // ── 0. Rebuild spatial grid — O(N) ───────────────────────────────────────
    grid_.rebuild(drones_);

    // ── 1. Decision pass — O(N·k) via grid ───────────────────────────────────
    decisions_ = decision_tree::decide_all(drones_, grid_);

    // ── 2. Apply decisions ────────────────────────────────────────────────────
    decision_tree::apply_decisions(drones_, std::move(decisions_));

    // ── 3. Move ───────────────────────────────────────────────────────────────
    move_drones();

    // ── 4. Combat ─────────────────────────────────────────────────────────────
    resolve_combat();
}

std::string Simulation::snapshot_json() const
{
    std::scoped_lock lock{mu_};
    return build_json();
}

// ── Private: movement ─────────────────────────────────────────────────────────

void Simulation::move_drones()
{
    for (auto& d : drones_) {
        if (!d.is_alive()) continue;

        Vec3 desired{};

        switch (d.state) {
        case DroneState::Attack:
            if (d.target_idx) {
                const auto& t = drones_[*d.target_idx];
                if (t.is_alive())
                    desired = (t.pos - d.pos).norm() * kMaxSpeed;
            }
            break;

        case DroneState::Flank:
            if (d.target_idx) {
                const auto& t = drones_[*d.target_idx];
                if (t.is_alive())
                    desired = ((t.pos - d.pos).norm()
                               + d.flank_offset * 1.2).norm() * kMaxSpeed;
            }
            break;

        case DroneState::Retreat: {
            const auto [eid, dummy] = grid_.nearest_enemy_any(d, drones_);
            if (eid != std::size_t(-1))
                desired = (d.pos - drones_[eid].pos).norm() * (kMaxSpeed * 1.1);
            break;
        }

        case DroneState::Patrol:
            if (d.pos.dist(d.waypoint) < 6.0)
                d.waypoint = random_waypoint();
            desired = (d.waypoint - d.pos).norm() * kPatrolSpeed;
            break;

        case DroneState::Dead:
            continue;
        }

        d.vel = d.vel.lerp(desired, kSteerFactor);
        if (const double spd = d.vel.len(); spd > kMaxSpeed)
            d.vel = d.vel * (kMaxSpeed / spd);
        d.pos = d.pos + d.vel * kTickDt;
        clamp_pos(d.pos);
    }
}

// ── Private: combat ───────────────────────────────────────────────────────────

void Simulation::resolve_combat()
{
    hits_.clear();

    for (const auto& d : drones_) {
        if (d.state != DroneState::Attack || !d.is_alive() || !d.target_idx)
            continue;
        const auto& t = drones_[*d.target_idx];
        if (t.is_alive() && d.pos.dist(t.pos) <= kAttackRange)
            hits_.push_back({*d.target_idx, kDamagePerTick});
    }

    for (const auto& [tidx, dmg] : hits_)
        drones_[tidx].health = std::max(0.0, drones_[tidx].health - dmg);
}

// ── Private: helpers ──────────────────────────────────────────────────────────

void Simulation::spawn_drones()
{
    drones_.clear();
    std::size_t id = 0;
    for (std::size_t i = 0; i < drones_per_team_; ++i)
        drones_.emplace_back(id++, Team::Alpha,
            Vec3{rand_range(-120,-80), rand_range(10,60), rand_range(-60,60)});
    for (std::size_t i = 0; i < drones_per_team_; ++i)
        drones_.emplace_back(id++, Team::Beta,
            Vec3{rand_range(80,120), rand_range(10,60), rand_range(-60,60)});
}

std::size_t Simulation::alive_count(Team team) const noexcept
{
    return static_cast<std::size_t>(
        std::ranges::count_if(drones_,
            [team](const Drone& d){ return d.team == team && d.is_alive(); }));
}

// ── Private: JSON serialisation (glaze) ───────────────────────────────────────

std::string Simulation::build_json() const
{
    const bool resetting = (phase_ == Phase::Resetting);

    WorldSnap snap;
    snap.tick       = tick_;
    snap.phase      = resetting ? "resetting" : "running";
    snap.alphaAlive = alive_count(Team::Alpha);
    snap.betaAlive  = alive_count(Team::Beta);
    snap.resetIn    = resetting ? std::max(0.0, reset_countdown_) : 0.0;

    snap.drones.resize(drones_.size());
    for (std::size_t i = 0; i < drones_.size(); ++i) {
        const auto& d = drones_[i];
        auto&       s = snap.drones[i];
        s.id     = static_cast<int>(d.id);
        s.team   = static_cast<int>(d.team);
        s.x = d.pos.x; s.y = d.pos.y; s.z = d.pos.z;
        s.vx = d.vel.x; s.vy = d.vel.y; s.vz = d.vel.z;
        s.health = std::max(0.0, d.health);
        s.maxHp  = d.max_health;
        s.state  = to_string(d.state);
    }

    json_buf_.clear();
    auto result = glz::write_json(snap, json_buf_);
    (void)result;
    return json_buf_;
}

File: src/sim/simulation.rs
use rand::Rng;
use serde::Serialize;

use super::decision_tree::{self, ATTACK_RANGE, DAMAGE_PER_TICK};
use super::drone::{Drone, DroneState, Team};
use super::vec3::Vec3;

// ── Constants ─────────────────────────────────────────────────────────────────

const TICK_DT:        f64 = 0.05;   // seconds per tick → 20 TPS
const MAX_SPEED:      f64 = 9.0;    // max drone speed  (units/s)
const PATROL_SPEED:   f64 = 4.5;    // patrol cruise speed
const BOUNDS_HALF:    f64 = 150.0;  // world ±X and ±Z
const BOUNDS_H_MAX:   f64 = 100.0;  // world max Y
const STEER_FACTOR:   f64 = 0.12;   // velocity smoothing (0..1)
const RESET_DELAY:    f64 = 5.0;    // seconds before respawn after one side is wiped

// ── Phase ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Phase { Running, Resetting }

// ── Simulation ────────────────────────────────────────────────────────────────

pub struct Simulation {
    pub drones:          Vec<Drone>,
    pub tick:            u64,
    pub phase:           Phase,
    pub drones_per_team: usize,
    reset_countdown:     f64,
}

impl Simulation {
    pub fn new(drones_per_team: usize) -> Self {
        let mut sim = Self {
            drones: Vec::new(),
            tick: 0,
            phase: Phase::Running,
            drones_per_team,
            reset_countdown: 0.0,
        };
        sim.spawn_drones();
        sim
    }

    // ── Tick ──────────────────────────────────────────────────────────────────

    pub fn step(&mut self) {
        self.tick += 1;

        // Check for end-of-battle
        if self.phase == Phase::Running {
            let alpha_alive = self.alive_count(Team::Alpha);
            let beta_alive  = self.alive_count(Team::Beta);
            if alpha_alive == 0 || beta_alive == 0 {
                self.phase          = Phase::Resetting;
                self.reset_countdown = RESET_DELAY;
            }
        }

        if self.phase == Phase::Resetting {
            self.reset_countdown -= TICK_DT;
            if self.reset_countdown <= 0.0 {
                self.spawn_drones();
                self.phase = Phase::Running;
            }
            return;
        }

        // ── 1. Decision pass (pure read, borrow-checker-safe) ─────────────────
        let decisions = decision_tree::decide_all(&self.drones);

        // ── 2. Apply decisions (mutate state fields) ──────────────────────────
        decision_tree::apply_decisions(&mut self.drones, decisions);

        // ── 3. Movement ───────────────────────────────────────────────────────
        // We need to read enemy positions for retreat, so collect retreat
        // targets first, then apply movement.
        let retreat_targets: Vec<Option<Vec3>> = self.drones.iter().map(|d| {
            if d.state == DroneState::Retreat {
                self.drones.iter()
                    .filter(|o| o.team != d.team && o.is_alive())
                    .map(|o| (o.pos, d.pos.dist_sq(o.pos)))
                    .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap())
                    .map(|(pos, _)| pos)
            } else {
                None
            }
        }).collect();

        let target_positions: Vec<Option<Vec3>> = self.drones.iter().map(|d| {
            d.target_idx.and_then(|idx| {
                self.drones.get(idx).filter(|t| t.is_alive()).map(|t| t.pos)
            })
        }).collect();

        let mut rng = rand::thread_rng();
        for (i, drone) in self.drones.iter_mut().enumerate() {
            if !drone.is_alive() { continue; }
            move_drone(drone, target_positions[i], retreat_targets[i], &mut rng);
        }

        // ── 4. Combat ─────────────────────────────────────────────────────────
        // Collect (attacker_idx, target_idx, damage) to avoid aliased &mut.
        let hits: Vec<(usize, usize, f64)> = self.drones.iter()
            .enumerate()
            .filter_map(|(i, d)| {
                if d.state != DroneState::Attack || !d.is_alive() { return None; }
                let tidx = d.target_idx?;
                let tpos = self.drones.get(tidx)?.pos;
                if d.pos.dist(tpos) <= ATTACK_RANGE {
                    Some((i, tidx, DAMAGE_PER_TICK))
                } else {
                    None
                }
            })
            .collect();

        for (_, tidx, dmg) in hits {
            if let Some(target) = self.drones.get_mut(tidx) {
                target.health = (target.health - dmg).max(0.0);
            }
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    fn spawn_drones(&mut self) {
        self.drones.clear();
        let mut rng = rand::thread_rng();
        let mut id  = 0usize;

        for _ in 0..self.drones_per_team {
            let pos = Vec3::new(
                -120.0 + rng.gen::<f64>() * 40.0,
                10.0   + rng.gen::<f64>() * 50.0,
                -60.0  + rng.gen::<f64>() * 120.0,
            );
            self.drones.push(Drone::new(id, Team::Alpha, pos));
            id += 1;
        }
        for _ in 0..self.drones_per_team {
            let pos = Vec3::new(
                80.0  + rng.gen::<f64>() * 40.0,
                10.0  + rng.gen::<f64>() * 50.0,
                -60.0 + rng.gen::<f64>() * 120.0,
            );
            self.drones.push(Drone::new(id, Team::Beta, pos));
            id += 1;
        }
    }

    pub fn alive_count(&self, team: Team) -> usize {
        self.drones.iter()
            .filter(|d| d.team == team && d.is_alive())
            .count()
    }

    // ── Snapshot ──────────────────────────────────────────────────────────────

    pub fn snapshot_json(&self) -> String {
        let snap = WorldSnap {
            tick:        self.tick,
            phase:       self.phase,
            alpha_alive: self.alive_count(Team::Alpha),
            beta_alive:  self.alive_count(Team::Beta),
            reset_in:    self.reset_countdown.max(0.0),
            drones:      self.drones.iter().map(DroneSnap::from_drone).collect(),
        };
        serde_json::to_string(&snap).expect("serialisation should never fail")
    }
}

// ── Movement ──────────────────────────────────────────────────────────────────

fn move_drone(
    d:            &mut Drone,
    target_pos:   Option<Vec3>,
    retreat_from: Option<Vec3>,
    rng:          &mut impl Rng,
) {
    let desired = match d.state {
        DroneState::Attack => {
            match target_pos {
                Some(tp) => tp.sub(d.pos).norm().scale(MAX_SPEED),
                None     => Vec3::default(),
            }
        }

        DroneState::Flank => {
            match target_pos {
                Some(tp) => {
                    let to_target = tp.sub(d.pos).norm();
                    let flank     = d.flank_offset;
                    to_target.add(flank.scale(1.2)).norm().scale(MAX_SPEED)
                }
                None => Vec3::default(),
            }
        }

        DroneState::Retreat => {
            match retreat_from {
                Some(enemy_pos) => d.pos.sub(enemy_pos).norm().scale(MAX_SPEED * 1.1),
                None            => Vec3::default(),
            }
        }

        DroneState::Patrol => {
            if d.pos.dist(d.waypoint) < 6.0 {
                d.waypoint = random_waypoint(rng);
            }
            d.waypoint.sub(d.pos).norm().scale(PATROL_SPEED)
        }

        DroneState::Dead => return,
    };

    // Exponential velocity smoothing toward desired
    d.vel = d.vel.lerp(desired, STEER_FACTOR);

    // Speed clamp
    let spd = d.vel.len();
    if spd > MAX_SPEED { d.vel = d.vel.scale(MAX_SPEED / spd); }

    d.pos = d.pos.add(d.vel.scale(TICK_DT));
    clamp_pos(&mut d.pos);
}

fn clamp_pos(v: &mut Vec3) {
    v.x = v.x.clamp(-BOUNDS_HALF, BOUNDS_HALF);
    v.y = v.y.clamp(2.0, BOUNDS_H_MAX);
    v.z = v.z.clamp(-BOUNDS_HALF, BOUNDS_HALF);
}

fn random_waypoint(rng: &mut impl Rng) -> Vec3 {
    Vec3::new(
        -100.0 + rng.gen::<f64>() * 200.0,
        10.0   + rng.gen::<f64>() * 60.0,
        -100.0 + rng.gen::<f64>() * 200.0,
    )
}

// ── Wire format (matches Go's WorldSnap / DroneSnap) ─────────────────────────

#[derive(Serialize)]
struct WorldSnap {
    tick:        u64,
    phase:       Phase,
    #[serde(rename = "alphaAlive")]
    alpha_alive: usize,
    #[serde(rename = "betaAlive")]
    beta_alive:  usize,
    #[serde(rename = "resetIn")]
    reset_in:    f64,
    drones:      Vec<DroneSnap>,
}

#[derive(Serialize)]
struct DroneSnap {
    id:     usize,
    team:   u8,
    x:      f64, y: f64, z: f64,
    vx:     f64, vy: f64, vz: f64,
    health: f64,
    #[serde(rename = "maxHp")]
    max_hp: f64,
    state:  DroneState,
}

impl DroneSnap {
    fn from_drone(d: &Drone) -> Self {
        Self {
            id:     d.id,
            team:   d.team as u8,
            x:      d.pos.x, y: d.pos.y, z: d.pos.z,
            vx:     d.vel.x, vy: d.vel.y, vz: d.vel.z,
            health: d.health.max(0.0),
            max_hp: d.max_health,
            state:  d.state,
        }
    }
}
